File: backend/api/passcode.py
import time
from backend.core import tg_client

class PasscodeHandler:
    def __init__(self, bridge):
        self.bridge = bridge

    async def has_passcode(self):
        await self.bridge._ensure_client()
        from backend.core import has_passcode_on_telegram
        return {"has_passcode": await has_passcode_on_telegram(tg_client.client)}

    async def set_passcode(self, passcode):
        await self.bridge._ensure_client()
        from backend.core import set_passcode_on_telegram, validate_passcode
        
        if not validate_passcode(passcode):
            return {"success": False, "error": "Passcode must be exactly 6 digits"}
        
        try:
            await set_passcode_on_telegram(tg_client.client, passcode)
            self.bridge._session_passcode = passcode
            return {"success": True}
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def verify_passcode(self, passcode):
        # Check if locked out
        if self.bridge._passcode_lockout_until and time.time() < self.bridge._passcode_lockout_until:
            retry_after = int(self.bridge._passcode_lockout_until - time.time())
            return {
                "valid": False,
                "error": "locked_out",
                "message": f"Too many failed attempts. Try again in {retry_after}s",
                "retry_after": retry_after
            }
        
        await self.bridge._ensure_client()
        from backend.core import verify_passcode_from_telegram
        
        valid = await verify_passcode_from_telegram(tg_client.client, passcode)
        
        if valid:
            self.bridge._failed_passcode_attempts = 0
            self.bridge._passcode_lockout_until = None
            self.bridge._session_passcode = passcode
            return {"valid": True}
        else:
            self.bridge._failed_passcode_attempts += 1
            attempts_remaining = 5 - self.bridge._failed_passcode_attempts
            
            if self.bridge._failed_passcode_attempts >= 5:
                self.bridge._passcode_lockout_until = time.time() + 30
                return {
                    "valid": False,
                    "error": "too_many_attempts",
                    "message": "Too many failed attempts. Locked for 30 seconds",
                    "locked_for": 30,
                    "attempts_remaining": 0
                }
            else:
                return {
                    "valid": False,
                    "error": "incorrect",
                    "message": f"Incorrect passcode. {attempts_remaining} attempts remaining",
                    "attempts_remaining": attempts_remaining
                }

    async def change_passcode(self, old_passcode, new_passcode):
        await self.bridge._ensure_client()
        from backend.core import change_passcode_on_telegram, validate_passcode
        
        if not validate_passcode(new_passcode):
            return {"error": "New passcode must be exactly 6 digits"}
        
        try:
            success = await change_passcode_on_telegram(tg_client.client, old_passcode, new_passcode)
            if success:
                self.bridge._session_passcode = new_passcode
                
                # Re-encrypt all V2 metadata messages
                try:
                    from backend.core import MetadataManager
                    print("PasscodeHandler: Starting metadata re-encryption...")
                    count = 0
                    async for msg in tg_client.client.iter_messages("me"):
                        if msg.text and msg.text.startswith("METADATA_V2_ENCRYPTED"):
                            try:
                                encrypted_str = msg.text.split("\n", 1)[1]
                                metadata = MetadataManager.from_json_encrypted(encrypted_str, old_passcode)
                                new_content = MetadataManager.to_json_encrypted(metadata, new_passcode)
                                await msg.edit(f"METADATA_V2_ENCRYPTED\n{new_content}")
                                count += 1
                            except Exception as e:
                                print(f"PasscodeHandler: Failed to re-encrypt message {msg.id}: {e}")
                    print(f"PasscodeHandler: Re-encryption complete. Processed {count} files.")
                except Exception as e:
                    print(f"PasscodeHandler: Critical error during re-encryption: {e}")
                
                return {"success": True}
            else:
                return {"error": "Incorrect old passcode"}
        except ValueError as e:
            return {"error": str(e)}

    async def reset_encryption(self):
        await self.bridge._ensure_client()
        from backend.core import reset_all_encrypted_data
        result = await reset_all_encrypted_data(tg_client.client)
        
        if hasattr(self.bridge, '_session_passcode'):
            self.bridge._session_passcode = None
        self.bridge._failed_passcode_attempts = 0
        self.bridge._passcode_lockout_until = None
        
        return {
            "success": True,
            "passcode_deleted": result["passcode_deleted"],
            "encrypted_files_deleted": result["encrypted_files_deleted"],
            "chunks_deleted": result.get("chunks_deleted", 0)
        }

File: frontend/src/stores/passcode.js
import { writable } from 'svelte/store';
import { hasPasscode, verifyPasscode, setPasscode, resetEncryption } from '../lib/api';

function createPasscodeStore() {
    const { subscribe, set, update } = writable({
        hasPasscode: false,
        isVerified: false,
        isLoading: false,
        error: null,
        attemptsRemaining: 5,
        lockedUntil: null
    });

    return {
        subscribe,

        async check() {
            update(state => ({ ...state, isLoading: true, error: null }));
            try {
                const result = await hasPasscode();
                update(state => ({
                    ...state,
                    hasPasscode: result.has_passcode,
                    isLoading: false
                }));
            } catch (error) {
                update(state => ({
                    ...state,
                    error: error.message,
                    isLoading: false
                }));
            }
        },

        async setup(passcode) {
            update(state => ({ ...state, isLoading: true, error: null }));
            try {
                const result = await setPasscode(passcode);
                if (result.success) {
                    update(state => ({
                        ...state,
                        hasPasscode: true,
                        isVerified: true,
                        isLoading: false
                    }));
                    return { success: true };
                } else {
                    update(state => ({
                        ...state,
                        error: result.error,
                        isLoading: false
                    }));
                    return { success: false, error: result.error };
                }
            } catch (error) {
                update(state => ({
                    ...state,
                    error: error.message,
                    isLoading: false
                }));
                return { success: false, error: error.message };
            }
        },

        async verify(passcode) {
            update(state => ({ ...state, isLoading: true, error: null }));
            try {
                const result = await verifyPasscode(passcode);

                if (result.valid) {
                    update(state => ({
                        ...state,
                        isVerified: true,
                        isLoading: false,
                        attemptsRemaining: 5,
                        lockedUntil: null,
                        error: null
                    }));
                    return { success: true };
                } else if (result.error === 'locked_out') {
                    const lockedUntil = Date.now() + (result.retry_after * 1000);
                    update(state => ({
                        ...state,
                        isLoading: false,
                        lockedUntil,
                        error: result.message
                    }));
                    return { success: false, locked: true, retryAfter: result.retry_after };
                } else if (result.error === 'too_many_attempts') {
                    const lockedUntil = Date.now() + (result.locked_for * 1000);
                    update(state => ({
                        ...state,
                        isLoading: false,
                        lockedUntil,
                        attemptsRemaining: 0,
                        error: result.message
                    }));
                    return { success: false, locked: true, lockedFor: result.locked_for };
                } else {
                    update(state => ({
                        ...state,
                        isLoading: false,
                        attemptsRemaining: result.attempts_remaining || 0,
                        error: result.message
                    }));
                    return {
                        success: false,
                        attemptsRemaining: result.attempts_remaining
                    };
                }
            } catch (error) {
                update(state => ({
                    ...state,
                    error: error.message,
                    isLoading: false
                }));
                return { success: false, error: error.message };
            }
        },

        async reset() {
            update(state => ({ ...state, isLoading: true, error: null }));
            try {
                const result = await resetEncryption();
                if (result.success) {
                    update(state => ({
                        hasPasscode: false,
                        isVerified: false,
                        isLoading: false,
                        error: null,
                        attemptsRemaining: 5,
                        lockedUntil: null
                    }));
                    return {
                        success: true,
                        passcodeDeleted: result.passcode_deleted,
                        filesDeleted: result.encrypted_files_deleted
                    };
                }
            } catch (error) {
                update(state => ({
                    ...state,
                    error: error.message,
                    isLoading: false
                }));
                return { success: false, error: error.message };
            }
        },

        async change(oldPasscode, newPasscode) {
            update(state => ({ ...state, isLoading: true, error: null }));
            try {
                // API call
                const { changePasscode } = await import('../lib/api');
                const result = await changePasscode(oldPasscode, newPasscode);

                if (result.success) {
                    update(state => ({
                        ...state,
                        isLoading: false,
                        error: null
                    }));
                    return { success: true };
                } else {
                    update(state => ({
                        ...state,
                        isLoading: false,
                        error: result.error
                    }));
                    return { success: false, error: result.error };
                }
            } catch (error) {
                update(state => ({
                    ...state,
                    isLoading: false,
                    error: error.message
                }));
                return { success: false, error: error.message };
            }
        },

        clearError() {
            update(state => ({ ...state, error: null }));
        },

        skip() {
            // User skips passcode setup, continues with plaintext
            update(state => ({
                ...state,
                hasPasscode: false,
                isVerified: true  // Allow access to app
            }));
        },

        resetState() {
            set({
                hasPasscode: false,
                isVerified: false,
                isLoading: false,
                error: null,
                attemptsRemaining: 5,
                lockedUntil: null
            });
        }
    };
}

export const passcode = createPasscodeStore();
